ci, iwyu: Treat warnings as errors for specific directories

hebasto opened the pull request:
This PR is the first step towards treating IWYU warnings as errors. At this stage, it applies only to the `crypto` and `index` directories.

hebasto commented:
Rebased.

hebasto commented on ci/test/03_test_script.sh:
```diff
@@ -180,6 +181,13 @@ if [ "${RUN_TIDY}" = "true" ]; then
+  TARGETS_WITH_FORCED_IWYU="bitcoin_clientversion bitcoin_crypto"
+  cd "${BASE_BUILD_DIR}"
+  bash -c "cmake -S ${BASE_ROOT_DIR} -DCMAKE_CXX_INCLUDE_WHAT_YOU_USE=\"include-what-you-use;-Xiwyu;--cxx17ns;-Xiwyu;--mapping_file=${BASE_ROOT_DIR}/contrib/devtools/iwyu/bitcoin.core.imp;-Xiwyu;--error\""
+  bash -c "cmake --build . ${MAKEJOBS} --clean-first --target ${TARGETS_WITH_FORCED_IWYU}"
```
> Could make sense to run this before the `tee /tmp/iwyu_ci.out` command, given that it errors?

This approach allows a developer to check other includes, not only the first one that fails.

> Maybe a `-- --keep-going` could be used for a slightly better output, if stuff is possibly missing?

I'm not sure what criteria you are using to define "better" in this case. The current branch adds one more config/build round details to the CI log and the following:
1. A message like the one posted in the PR description, if there is a warning.
2. Nothing, if there are no warnings.

hebasto replied:
> I think the message you posted in the PR description only shows a single error in a single dependency. I am wondering what happens if there is more than one error in different targets, possibly ones where cmake will never get to print the second error.

[Here](https://api.cirrus-ci.com/v1/task/5614699137466368/logs/ci.log) multiple error are reported.

hebasto replied:
1. The [change](https://github.com/bitcoin/bitcoin/compare/16c69e927cccc1e93beaa10c621afd6a8f422e6c..0c3c7d07599d0086d5151949a69ffc12b6166d19) touches 8 files. The modification in `src/crypto/siphash.h` is a no-op from IWYU's perspective.

2. CMake integrates the IWYU into the compilation step. Therefore, it reports an error for each object file individually.
However, the diagnostic output includes both headers and the source file:
```
[12:32:50.757] Warning: include-what-you-use reported diagnostics:
[12:32:50.757] 
[12:32:50.757] /ci_container_base/src/crypto/poly1305.h should add these lines:
[12:32:50.757] #include <span>      // for span
[12:32:50.757] 
[12:32:50.757] /ci_container_base/src/crypto/poly1305.h should remove these lines:
[12:32:50.757] 
[12:32:50.757] The full include-list for /ci_container_base/src/crypto/poly1305.h:
[12:32:50.757] #include <span.h>    // for UCharCast
[12:32:50.757] #include <stdint.h>  // for uint32_t
[12:32:50.757] #include <cassert>   // for assert
[12:32:50.757] #include <cstddef>   // for byte, size_t
[12:32:50.757] #include <span>      // for span
[12:32:50.757] ---
[12:32:50.757] 
[12:32:50.757] (/ci_container_base/src/crypto/poly1305.cpp has correct #includes/fwd-decls)
[12:32:50.757] 
[12:32:50.757] gmake[3]: *** [src/crypto/CMakeFiles/bitcoin_crypto.dir/build.make:188: src/crypto/CMakeFiles/bitcoin_crypto.dir/poly1305.cpp.o] Error 1
```

I can see all [changes](https://github.com/bitcoin/bitcoin/compare/16c69e927cccc1e93beaa10c621afd6a8f422e6c..0c3c7d07599d0086d5151949a69ffc12b6166d19), except the unrelated one, in the IWYU diagnostics. Am I missing anything in particular?

hebasto replied:
> Not sure if this is a valid concern, but an alternative approach would be to just call `git diff` below on the `TARGETS_WITH_FORCED_IWYU` folders (not targets). Then exit, when the diff is not empty?

Thanks! [Reworked](https://github.com/bitcoin/bitcoin/pull/31308#issuecomment-2920042768).

maflcko reviewed (commented):
left some comments

fanquake commented:
> While this approach introduces some duplication at present, it will ultimately enable us to consolidate how the IWYU tool is executed, allowing for a choice between using the iwyu_tool.py script or CMake's built-in functionality.

Reading the PR description, it's not really clear why we'd need to retain using both tools? Either the CMake built-in should be better in some way, or at least equivalent in functionality, in which case we should just switch to using it?

fanquake commented on contrib/devtools/iwyu/bitcoin.core.imp:
```diff
@@ -1,3 +1,11 @@
-# Nothing for now.
 [
+  # Compiler intrinsics
+  { include: [ "<emmintrin.h>", private, "<immintrin.h>", public ] },
```
Same here?

fanquake replied:
> they do not fall within the scope of the IWYU repository.

IWYU has mappings for intrinsics though?
 - https://github.com/include-what-you-use/include-what-you-use/blob/master/gcc-8.intrinsics.imp
 - https://github.com/include-what-you-use/include-what-you-use/blob/master/clang-6.intrinsics.imp

hebasto replied:
> Same here?

Please see https://github.com/include-what-you-use/include-what-you-use/issues/1764.

maflcko reviewed (commented):
review ACK e9143a68f34f350d7fa4405389222da4dfa9a394 🎰

<details><summary>Show signature</summary>

Signature:

```
untrusted comment: signature from minisign secret key on empty file; verify via: minisign -Vm "${path_to_any_empty_file}" -P RWTRmVTMeKV5noAMqVlsMugDDCyyTSbA3Re5AkUrhvLVln0tSaFWglOw -x "${path_to_this_whole_four_line_signature_blob}"
RUTRmVTMeKV5npGrKx1nqXCw5zeVHdtdYURB/KlyA/LMFgpNCs+SkW9a8N95d+U4AP1RJMi+krxU1A3Yux4bpwZNLvVBKy0wLgM=
trusted comment: review ACK e9143a68f34f350d7fa4405389222da4dfa9a394 🎰
fi/hWQQqCAdnfz0F4yhi8AMYjYOl79DanoxZsC6Am2gCztYYNf/izs+yQjc7oCt14KzsBJkAcSvQN6BVRVniAQ==
```

</details>

fanquake commented on ci/test/03_test_script.sh:
```diff
@@ -33,7 +33,9 @@ echo "=== BEGIN env ==="
 
-(
+if [ "$RUN_TIDY" != "true" ]; then
+  # The tidy task has no UB detector, so skip the patch there to avoid issues
```
> to avoid issues

What issues? Do I understand correctly, that if we land the patch in the subtree, this will just break again?

hebasto replied:
The following `git diff` prints this patch even IWYU did not apply any modifications.

hebasto replied:
> Doesn't matter much, but the comment could also say "... so skip the patch because it is not needed and to avoid issues ...".

Thanks! [Adjusted](https://github.com/bitcoin/bitcoin/pull/31308#issuecomment-2936303614).

fanquake commented on contrib/devtools/iwyu/bitcoin.core.imp:
```diff
@@ -1,3 +1,11 @@
+  # False positive warnings
+  # See: https://github.com/include-what-you-use/include-what-you-use/issues/1616
+  { include: [ "<iterator>", public, "<utility>", public ] },
+  { include: [ "<tuple>", public, "<array>", public ] },
```
?

hebasto replied:
> Is there an issue open for this upstream?

Please see https://github.com/include-what-you-use/include-what-you-use/issues/1763.

hebasto replied:
[Reworked](https://github.com/bitcoin/bitcoin/pull/31308#issuecomment-2936303614) into a local pragma.

hebasto commented on src/crypto/hmac_sha512.cpp:
```diff
@@ -4,6 +4,7 @@
 #include <crypto/hmac_sha512.h>
 
+#include <crypto/sha512.h>
 #include <string.h>
```
Thanks! Fixed.

hebasto commented on ci/test/03_test_script.sh:
```diff
@@ -172,13 +172,27 @@ if [ "${RUN_TIDY}" = "true" ]; then
+  }
+
+  run_iwyu "compile_commands_iwyu_errors.json"
+  git diff --quiet -- ':(exclude)src/leveldb/db/db_impl.cc' || { git --no-pager diff -- ':(exclude)src/leveldb/db/db_impl.cc'; exit 1; }
```
Thanks! Reworked.

jonatack reviewed (commented):
Concept ACK

hebasto commented on ci/test/03_test_script.sh:
```diff
@@ -178,9 +178,8 @@ if [ "${RUN_TIDY}" = "true" ]; then
            -- -Xiwyu --cxx17ns -Xiwyu --mapping_file="${BASE_ROOT_DIR}/contrib/devtools/iwyu/bitcoin.core.imp" \
            -Xiwyu --max_line_length=160 \
            2>&1 | tee /tmp/iwyu_ci.out
-  cd "${BASE_ROOT_DIR}/src"
```
I've tested the pre-CMake master branch @ 80f00cafdeef0600fa1a5e906686720786c2336c, and this line wasn't necessary there.

hebasto commented:
Rebased.

All feedback from @maflcko and @fanquake has been addressed.

The IWYU-related workarounds are now documented with references to the relevant upstream issues.

hebasto commented:
Rebased.

@maflcko 

Thank you for the review! Your feedback has been addressed.

hebasto commented on src/index/txindex.cpp:
```diff
@@ -4,13 +4,31 @@
+#include <cassert>
+#include <cstdint>
+#include <exception>
+#include <fcntl.h>
```
Sure. [Reworked](https://github.com/bitcoin/bitcoin/pull/31308#issuecomment-3062529488).

hebasto commented on src/crypto/hex_base.cpp:
```diff
@@ -5,8 +5,11 @@
 #include <cstring>
 #include <string>
+// A workaround for https://github.com/include-what-you-use/include-what-you-use/issues/1763.
+// IWYU pragma: no_include <tuple>
```
Thanks! [Updated](https://github.com/bitcoin/bitcoin/pull/31308#issuecomment-3062529488).

fanquake reviewed (changes requested):
This makes the CI dependant on being run on x86_64; the tidy job will no-longer pass when run on aarch64.

fanquake commented:
> may be good to fix the aarch64 ci.

How would you fix it, given IWYU is giving different recommendations, based on the architecture?

ryanofsky commented:
> How would you fix it, given IWYU is giving different recommendations, based on the architecture?

Easiest way is to use `// IWYU pragma: keep` when IWYU incorrectly tells you to remove an include that should be kept and `// IWYU pragma: no_include` when it tells you to add an include that shouldn't be added. The pragmas will suppress errors on whatever platform is complaining and be ignored on other platforms. I dealt with this recently in https://github.com/bitcoin-core/libmultiprocess/commit/0d5f1faae5da26205ef9563cb9cc157725ab9479

ryanofsky commented:
> The correct fix is to just add the includes iwyu is asking for.

I think this is about cases where IWYU on one platform asks you to add (or remove) an include, and IWYU on other platform asks you to remove (or add) the same include. A reasonable way to resolve these is to decide which IWYU is correct and add pragmas to silence the other IWYU. Other ways to resolve it could be to define mappings, or submit fixes upstream, or choose a single platform to be your canonical IWYU platform and ignore other platforms.

I think pragmas are helpful because they make it easier to enable IWYU more places and allow developers to run it locally, but they do have downsides because they can become incorrect over time and be a maintenance burden. Many approaches seem reasonable and I don't have a recommendation, this was just a suggestion.

hebasto committed:
refactor: Fix includes in `crypto` directory

IWYU issue #1763 appears to be a corner case, so it has been addressed
using a local pragma rather than a global mapping.

hebasto commented:
> may be good to fix the aarch64 ci.

Fixed.

fanquake commented on contrib/devtools/iwyu/bitcoin.core.imp:
```diff
@@ -1,3 +1,16 @@
+  { "include": [ "<smmintrin.h>", "private", "<immintrin.h>", "public" ] },
+  { "include": [ "<tmmintrin.h>", "private", "<immintrin.h>", "public" ] },
+
+  # libc symbols.
```
Has this been upstreamed / why is it needed?

hebasto commented:
Rebased to resolve a conflict with the merged #33116.

hebasto commented:
Rebased and addressed @maflcko's [comment](https://github.com/bitcoin/bitcoin/pull/31308#discussion_r2313401875).

hebasto commented on ci/test/03_test_script.sh:
```diff
@@ -199,14 +201,27 @@ if [ "${RUN_TIDY}" = "true" ]; then
+  }
+
+  run_iwyu "compile_commands_iwyu_errors.json"
+  git --no-pager diff --exit-code
```
Thanks! [Done](https://github.com/bitcoin/bitcoin/pull/31308#issuecomment-3241834149).

l0rinc reviewed (commented):
Concept ACK

ryanofsky commented on ci/test/03_test_script.sh:
```diff
@@ -33,7 +33,9 @@ echo "=== BEGIN env ==="
-(
+if [ "$RUN_TIDY" != "true" ]; then
+  # The tidy task has no UB detector, so skip the patch because it is not
+  # needed and to avoid issues later on in the task.
```
In commit "ci: Do not patch `leveldb` to workaround UB in "tidy" CI job" (b495e2dc95db42e3219720f5beeb6c136d067a4a)

Could the comment clarify what the issues are? Even just a hint or pointer to more information would be helpful. if they are IWYU issues, maybe just say "to avoid IWYU issues"

ryanofsky commented on ci/test/03_test_script.sh:
```diff
@@ -201,14 +201,30 @@ if [ "${RUN_TIDY}" = "true" ]; then
 
+  # TODO: Consider enforcing IWYU across the entire codebase.
+  FILES_WITH_FORCED_IWYU="/src/(crypto|index)/.*\\.cpp"
+  jq --arg patterns "$FILES_WITH_FORCED_IWYU" 'map(select(.file | test($patterns)))' "${BASE_BUILD_DIR}/compile_commands.json" > "${BASE_BUILD_DIR}/compile_commands_iwyu_errors.json"
```
In commit "ci, iwyu: Treat warnings as errors for specific directories" (1d49346465de8edec85bb2d498859b4fbc346935)

I think it'd be a little better to called this ENFORCED_IWYU instead of FORCED_IWYU. Both make sense but former seems a little clearer.

ryanofsky reviewed (approved):
Code review ACK 1d49346465de8edec85bb2d498859b4fbc346935. Makes sense to try IWYU in a few directories like this to see how stable it is and make sure it won't cause problems.

hebasto committed:
ci, iwyu: Treat warnings as errors for specific directories

Currently, this applies only to the `crypto` and `index` directories.

hebasto commented:
The feedback from @ryanofsky has been addressed.

ryanofsky reviewed (approved):
Code review ACK d03d861cb6d81f90f3b4add2db3dbdc0bf95a0ef. Just fixing up comment and changing variable name since last review. Thanks for the updates!

hebasto committed:
ci: Do not patch `leveldb` to workaround UB in "tidy" CI job

hebasto committed:
refactor: Fix includes in `index` directory

ryanofsky reviewed (approved):
Code review ACK 02d2b5a11c921ef71c971ee80eb3dfbc75c8cb0d. Just rebased and update tidy patch comment again since last review

maflcko commented:
rfm?

willcl-ark reviewed (approved):
ACK 02d2b5a11c921ef71c971ee80eb3dfbc75c8cb0d

I like the approach here of enforcing as-we-go through the codebase. Something similar was done recently in `nixpkgs` [with their `treefmt` formatter](https://youtu.be/3m_cvSuOB-Y?si=GuuU3aC7IR_FKB6q).

Tested that this worked locally by modifying the test script and running.

<details>
<summary>Details</summary>

```bash
#!/usr/bin/env bash
set -x

BASE_ROOT_DIR="$HOME/src/core/bitcoin"
BASE_BUILD_DIR="$BASE_ROOT_DIR/build"
FILES_WITH_ENFORCED_IWYU="$HOME/src/core/bitcoin/src/(crypto|index)/.*\\.cpp"

jq --arg patterns "$FILES_WITH_ENFORCED_IWYU" 'map(select(.file | test($patterns)))' "${BASE_BUILD_DIR}/compile_commands.json" > "${BASE_BUILD_DIR}/compile_commands_iwyu_errors.json"
jq --arg patterns "$FILES_WITH_ENFORCED_IWYU" 'map(select(.file | test($patterns) | not))' "${BASE_BUILD_DIR}/compile_commands.json" > "${BASE_BUILD_DIR}/compile_commands_iwyu_warnings.json"

cd "${BASE_ROOT_DIR}" || exit

run_iwyu() {
    mv "${BASE_BUILD_DIR}/$1" "${BASE_BUILD_DIR}/compile_commands.json"
    "$(command -v iwyu_tool.py)" \
         -p "${BASE_BUILD_DIR}" "${MAKEJOBS}" \
         -- -Xiwyu --cxx17ns -Xiwyu --mapping_file="${BASE_ROOT_DIR}/contrib/devtools/iwyu/bitcoin.core.imp" \
         -Xiwyu --max_line_length=160 \
         2>&1 | tee /tmp/iwyu_ci.out
    "$(command -v fix_includes.py)" --nosafe_headers < /tmp/iwyu_ci.out
}

run_iwyu "compile_commands_iwyu_errors.json"
if ! ( git --no-pager diff --exit-code ); then
    echo "^^^ ⚠️ Failure generated from IWYU"
    false
fi

run_iwyu "compile_commands_iwyu_warnings.json"
git --no-pager diff
```

</details>

hebasto commented on contrib/devtools/iwyu/bitcoin.core.imp:
```diff
@@ -1,3 +1,16 @@
-# Nothing for now.
 [
+  # Compiler intrinsics.
+  # See: https://github.com/include-what-you-use/include-what-you-use/issues/1764.
+  { "include": [ "<emmintrin.h>", "private", "<immintrin.h>", "public" ] },
+  { "include": [ "<smmintrin.h>", "private", "<immintrin.h>", "public" ] },
+  { "include": [ "<tmmintrin.h>", "private", "<immintrin.h>", "public" ] },
```
This has been addressed upstream in https://github.com/include-what-you-use/include-what-you-use/commit/6a113bcf7017b8d5b7b4358c8a0fb16be96ad39a.
